### ISSUE TITLE
client sftp exists crash

### ISSUE THREAD

**Issue by oussamaABID** [theophilusx/ssh2-sftp-client#190]
when i call exists function with path folder not existing. i have error : 
```js
stack:"Error: sftp.exists: sftp.realPath: 
Cannot find message [/Inbox/tip/test] /Inbox/tip/test\n 
at formatError (\node_modules\ssh2-sftp-client\src\index.js:56:10)\n    
at SftpClient.exists (\node_modules\ssh2-sftp-client\src\index.js:241:27)\n    
at process._tickCallback (internal/process/next_tick.js:68:7)"
```
my code : 
```js
 clientSftp.connect({
        host: ssh2.host,
        port: ssh2.port,
        username: ssh2.username,
        password: ssh2.password
      }) .then(() => {
        const folder = `${ssh2.path}/${type}`;
        return clientSftp.exists(folder);
      }) .then((exist) => {
          return resolve(exist);
      }).catch(err => {
          self.logger.error( `transport to ${localFile}/${remoteFile} error`+ err.message);
          return reject(err);
        });
```

**Comment by theophilusx** [comment]
Can you provide

- node version
- node platform (i.e. Linux, windows, macOS etc)
- ssh2-sftp-client version
- The actual value for the 'folder' variable. I suspect that this is such an unexpected path value that it is causing the module to crash, so I need to see the value to trace what is going on.

Also, is that code running inside a 'new Promise()' block? If not, you should not be calling resolve/reject.

**Comment by oussamaABID** [comment]
hi,

- node version : 10.16.3
- os: Windows10
- ssh2-sftp-client@4.1.0
- const folder = '/Inbox/tip/test'   _(Folders /Inbox and  /Inbox/tip exist in server)_
- yes i have Promise

my code : 
```js
  /**
   * send File By SSh2
   *
   * @param {client-sftp} clientSftp
   * @param {Object} ssh2
   * @param {String} localFile
   * @param {String} remoteFile
   * @param {String} subFolder
   * @returns
   */
  async _sendFileWithSsh2( clientSftp, ssh2, localFile, remoteFile, subFolder){
    const self = this;
    return new Promise((resolve, reject)=>{
      subFolder = subFolder.toLowerCase();
      let r;
      clientSftp.connect({
        host: ssh2.host,
        port: ssh2.port,
        username: ssh2.username,
        password: ssh2.password
      }) .then(() => {
        const folder = `${ssh2.path}/${subFolder}`;
        return clientSftp.exists(folder);
      })
        .then(exist => {
          if(exist === 'd'){ 
            self.logger.info( `folder : ${ssh2.path}/${subFolder} exist`);
            return 'exist';
          }else{
            self.logger.info( `create folder : ${ssh2.path}/${subFolder}`);
            const folder = `${ssh2.path}/${subFolder}`;
            return clientSftp.mkdir(folder, true);
          }      
      
        }).then(() => {
          self.logger.info( `start transport from ${localFile} to ${remoteFile}`);
          return clientSftp.put(localFile, remoteFile);
          
        }).then((res) => {
          self.logger.info( `end transport from  ${localFile} to ${remoteFile}`);
          r = res;
          return clientSftp.end();
        }).then(()=>{
          if(r)resolve({from: `${localFile}`, to:`${remoteFile}`});
          else reject({from: `${localFile}`, to:`${remoteFile}`, response:'empty'});
        })
        .catch(err => {
          self.logger.error( `transport to ${localFile}/${remoteFile} error`+ err.message);
          return reject(err);
        });
    });
  }
```

**Comment by theophilusx** [comment]
Thanks for the info. 

First, please upgrade to v4.2.3. There are a number of bug fixes, including one relating to exists() in the latest version which are not in the version you are using. Once upgraded, test again and let me know if you still get the same error.

**Comment by oussamaABID** [comment]
thanks for replay
i test with v4.2.3 . she not fix the probleme :(

**Comment by theophilusx** [comment]
I've not yet been able to reproduce the error you are seeing. Will continue to investigate. 

However, I notice that your code could be made a lot simpler. The call to new Promise is  unnecessary within the context of an async function. An async function always returns a promise. If no errors are generated in the function, the returned promise is resolved with whatever value you return. If an error is raised, then the function returns a rejected promise with the error as the value (though you can change this by wrapping everything in a try/catch). 

Also note that your better off using fastPut() rather than put(). The difference is that fastPut will do the upload in parallel, which will be faster. The fastPut/fastGet methods are far better when all you want to do is upload/download files. The get()/put() methods are useful whenh you want to do some additional processing of the data as part of the transfer (such as compressing or decompressing the files etc). 

So your code could be (untested)

```
  /**
   * send File By SSh2
   *
   * @param {client-sftp} clientSftp
   * @param {Object} ssh2
   * @param {String} localFile
   * @param {String} remoteFile
   * @param {String} subFolder
   * @returns
   */
  async _sendFileWithSsh2( clientSftp, ssh2, localFile, remoteFile, subFolder){
  const self = this;

  try {
    subFolder = subFolder.toLowerCase();
    await clientSftp.connect({
      host: ssh2.host,
      port: ssh2.port,
      username: ssh2.username,
      password: ssh2.password
    });
    const folder = `${ssh2.path}/${subFolder}`;
    let exist = await clientSftp.exists(folder);
    if (exist === 'd') {
      self.logger.info(`folder : ${ssh2.path}/${subFolder} exist`);
    } else {
      self.logger.info(`create folder : ${ssh2.path}/${subFolder}`);
      const folder = `${ssh2.path}/${subFolder}`;
    }
    self.logger.info(`start transport from ${localFile} to ${remoteFile}`);
    let res = await clientSftp.fastPut(localFile, remoteFile);
    self.logger.info(`end transport from  ${localFile} to ${remoteFile}`);
    return {
      from: localFile,
      to: remoteFile
    };
  } catch (err) {
    self.logger.error(`transport to ${localFile}/${remoteFile} error ${err.message}`);
   // just return a value. Alterntively, you could throw a new error if you want the returned value 
   // to be a rejected promise
    return {
      from: localFile,
      to: remoteFile,
      response: 'empty',
      error: err.message
    };
  } finally {
    await clientSftp.end();
  }
}
```

**Comment by theophilusx** [comment]
I have tried a number of different approaches, but cannot reproduce your error from either Linux or Windows 10. The error message you report is quite odd as well as there is nothing which would generate the error test of "Cannot find message".  The only way Ive been able to generate an error is when I try to do an exists() on a remote path which exists, but for which there is no permission for the current user. In this case, it generates an error which says Permisson denied". In all other cases, it just returns false or 'd' for a directory and '-' for a file. 

In the existscheck branch of the repository, there is a script in the examples directory called exists.js. Can you try running that and see if you get the same error. You will need to adjust the script to add your credentials etc. Run it as 'node exists.js /path/to/test

**Comment by theophilusx** [comment]
I have made some changes in the error-codes branch which might help with this issue. The main changes are to add consistent error codes to error objects, but I ahve also added an uncaughtException handler, which should stop any unexpected errors causing the process to abruptly exit. Give it a try and see if it helps. It is the error-codes branch.

**Comment by oussamaABID** [comment]
hi,
thanks for invesigated, and your code. I can t use putFast , it s MFT server by IBM he not support chunk file :(

I read the exist() implemntation code, and I try some tests:


with this code : 
var Client = require('ssh2').Client;
 
```js
var conn = new Client();
conn.on('ready', function() {
  console.log('Client :: ready');
  conn.sftp(function(err, sftp) {
    if (err) throw err;
    sftp.readdir('Inbox/tip/toto', function(err, list) { // Inbox/tip exist , toto no in server
      if (err) throw err;
      console.dir(list);
      conn.end();
    });
  });
}).connect({
  host: 'host',
  port: '22',
  username:'user',
  password:'pwd'
});
```
and the error is : 
```js
message:"User [APUSER] has no view rights to [/APBOX/Inbox/tip/toto]."
stack:"Error: User [APUSER] has no view rights to [/APBOX/Inbox/tip/toto].\n    at SFTPStream._transform (\node_modules\ssh2-streams\lib\sftp.js:412:27)\n    at SFTPStream.Transform._read (_stream_transform.js:190:10)\n    at SFTPStream._read (\micro_service\doc-gen-distribution\node_modules\ssh2-streams\lib\sftp.js:183:15)\n    at SFTPStream.Transform._write (_stream_transform.js:178:12)\n    at doWrite (_stream_writable.js:415:12)\n    at writeOrBuffer (_stream_writable.js:399:5)\n    at SFTPStream.Writable.write (_stream_writable.js:299:11)\n    at Channel.ondata (_stream_readable.js:709:20)\n    at Channel.emit (events.js:198:13)\n    at addChunk (_stream_readable.js:288:12)"
__proto__:Object {constructor: , name: "Error", message: "", …}
```

**Comment by theophilusx** [comment]
OK, so we know the error is being generated by the ssh2 module and we know the reason is because the user you are connecting with does not have correct permissions to read the target directory. 

The only remaining question is why the ssh2-sftp-client module is not catching that error correctly, resulting in the error bubbling up and causing the node process to exit. I need you to try two things

1. See what output you get if you run the script examples/exists.js in the repository. Post the output from running that script. (you will need to modify it slightly). 

2. Try the version in the error-codes branch of the repository. I made some changes today, including adding a listener for uncaughtExceptions which might resolve the issue. My suspicion is that for some reason, the error from the remote sftp server is not being propagated by the ssh2 module in the normal manner, so it is not returned in the error handler or raised as an error event. This would result in ssh2-sftp-client not catching it and it bubbling  up to kill the node process. The uncaughtException handler should stop this.

**Comment by oussamaABID** [comment]
Thanks I will do that tonight 
When you wrote that : 

> the user you are connecting with does not have correct permissions to read the target directory.

It's not entirely correct,  because I don't get an error when i create the toto folder in the server.
In my test '/Inbox/tip/toto'
Only Inbox and tip exist.

**Comment by theophilusx** [comment]
The error from ssh2 is a permission error. It is possible to have permission to create a directory and not have permission to list the contents of that directories parent(s). On *nix systems, you need both the execute and the read permissions in order to list a directory. If you only have the execute permission and you have write permission, you will be able to create a sub directory, but you won't be able to list the parent directory where you created the sub directory. The ability to create a directory is not a guarantee you have the permission to view (readdir). 

Note also that the ssh2-sftp-client module walks the directory tree when checking for the existence of a file/directory. If any of the directories in the path do not have both read and execute permissions, a permission error will be raised.  

The updated version in the error-codes branch should fix the issue of the node process exiting due to an uncaught error and allow it to be caught by your code. It will not do anything about the permission error itself as this is an error being raised by the ssh2 module. It is very unlikely this is an incorrect error, but if it is, you will need to raise that as an issue with the ssh2 maintainer as there is nothing I can do about errors generated by ssh2 other than propogate them up to the client code..

**Comment by oussamaABID** [comment]
Ok I will test  the error branch tonight .

I have the permission to list content folder.
The readdir callback when the folder exists is correct.

**Comment by oussamaABID** [comment]
hi i just doing  my tests, and i thing this line `if (err.message.match(/No such file/)) ` no catch the error.
The real error is generated by  `let absPath = await this.realPath(remotePath); `
Tofix this i thing we need other catch.

```js
/**
 * @async

 * Tests to see if an object exists. If it does, return the type of that object
 * (in the format returned by list). If it does not exist, return false.
 *
 * @param {string} path - path to the object on the sftp server.
 *
 * @return {boolean} returns false if object does not exist. Returns type of
 *                   object if it does
 */
SftpClient.prototype.exists = async function(remotePath) {
  try {
    if (!this.sftp) {
      throw formatError('No SFTP connection available', 'sftp.exists');
    }
    let absPath = await this.realPath(remotePath); // hi crash here 
    return this._exists(absPath);
  } catch (err) {
    if (err.message.match(/No such file/)) {
      return false;
    }
    throw formatError(err, 'sftp.exists');
  }
};
```
![Capture d’écran (5)](https://user-images.githubusercontent.com/7701298/69504090-2ab91780-0eee-11ea-913f-008544c2825d.png)

**Comment by oussamaABID** [comment]
Succes

<img width="597" alt="Success" src="https://user-images.githubusercontent.com/7701298/69504227-de220c00-0eee-11ea-81b0-bbc951a04829.PNG">

Error: 

<img width="623" alt="failed" src="https://user-images.githubusercontent.com/7701298/69504233-ea0dce00-0eee-11ea-8da3-fc1b59aa8072.PNG">

**Comment by theophilusx** [comment]
Odd, I added a comment about fixing this 5 days ago and now it is not there! 

Anyway, I think your analysis was correct and I have fixed the catch test to use the error code rather than the message, so this issue should now be resolved. This change to add error.code properties in a consistent manner was already undeway (since node v10 improved this situation). 

I have released a new version (v4.3.0) which has these fixes included.
